h2o-core/src/main/java/water/rapids/BinaryMerge.java

```diff
@@ -357,145 +351,134 @@ private long bsearchLeft(long x, boolean returnLow, long upp)  {
+   * we will try to find match between the left frame and rite frame and include those rows in the final merged frame.
+   * If no match is found between the two frames but allLeft = true, we will go ahead and just include all the rows
+   * having the same MSB in the left frame into the final merged frame.
+   * 
+   * If allLeft is true, we will always iterate over the leftFrame keys and use binary search over the rite frame keys.
+   * We will count all duplicate keys in the left and rite frames as well.  During the search process, we should be 
+   * able to shrink the range of rite frame keys to search over as we gain more knowledge over the relative key 
+   * sizes between the two frames and all keys are sorted already.
+   * 
+   * However, when allLeft is false, we will iterate over the frames that has the smallest number of keys instead. 
+   * Hence, in this case, we can search over either the left or rite frame depending on the number of rows to be 
+   * searched through.  
+   * 
+   * @param leftLowIn: left frame lowest row number minus 1
+   * @param leftUppIn: number of rows in left frame with specific MSB
+   * @param riteLowIn: rite frame lowest row number minus 1 with specific MSB
+   * @param riteUppIn: number of rows in rite frame with specific MSB
+   *                  
+   */
+  private void binaryMergeOneMSB(long leftLowIn, long leftUppIn, long riteLowIn, long riteUppIn) {
+    if (!_allLeft && riteUppIn==0) return;  // no merging possible with empty rite frame here
+    boolean leftFrameIterate = _allLeft?true:((leftUppIn-leftLowIn)>(riteUppIn-riteLowIn)?false:true);
+    long iterIndex = leftFrameIterate?leftLowIn+1:riteLowIn+1;
+    long iterUppIn = leftFrameIterate?leftUppIn:riteUppIn;
+    long binarySearchUppIn = leftFrameIterate?riteUppIn:leftUppIn;
+    long iterHigh, binarySearchLow, binarySearchHigh, tempValue;  // temp variables to perform search
+    long[] midSearchInd = new long[3]; // store match row, riteLow, riteUpp
+    midSearchInd[1] = leftFrameIterate?riteLowIn:leftLowIn;
+    midSearchInd[2] = leftFrameIterate?riteUppIn:leftUppIn;
+    int iterLen=0, binarySearchLen=0;  // number of iterate Frame/binarySearch frame rows to be included to merged frame
+    
+    while (iterIndex < iterUppIn) { // for each left row, find matches in the right frame using binary search
+      midSearchInd = bsearchRiteMatch(iterIndex, midSearchInd, leftFrameIterate);  // try to find match for leftFrame at leftIndex in riteFrame
+      // extend index to include/skip over duplicates in iterate frame
+      iterHigh = iterIndex+1;
+      while (iterHigh<iterUppIn && frameKeyEqual(leftFrameIterate?_leftKO._key:_riteKO._key, iterIndex, iterHigh,
+              leftFrameIterate?_leftKO:_riteKO, leftFrameIterate?_leftSB:_riteSB)) {
+        iterHigh++; // find iterate frame duplicates
+      }
+      iterLen = (int) (iterHigh-iterIndex); // number of iterate frame rows to be included
```

Comment:
it seems to me that iterlen is unnecessary and be replaced below with iterHigh-iterIndex

Reply:
True.  fixed.